### DISCUSSION TITLE
What does d-separation on non-disjoint sets mean?

### DISCUSSION THREAD

**Discussion by galenseilis** [networkx/networkx#6889] [Q&A] [answered]
I'm interested in writing a function to parse Lavaan-like strings used in [semopy](https://semopy.com/) into networkx, applying d-operators to get a new graph, converting the graph back to a Lavaan-like string, and then constructing the appropriate structural equation model from that. That's the background context of this question.

The foreground of this question is that I noticed that `nx.d_separated` allows us to check non-disjoint sets. I've created an example DAG based on this figure from [Bisset *et al* 2023](https://www.frontiersin.org/articles/10.3389/fmicb.2023.1097909/full):

![](https://www.frontiersin.org/files/Articles/1097909/fmicb-14-1097909-HTML/image_m/fmicb-14-1097909-g001.jpg)

Here is a minimum working example of that SEM converted to a `nx.DiGraph` object, and then checked for d-separation on non-disjoint sets. This script should return `True` for this check of d-separation using `nx.d_separated`.

```python
import networkx as nx

# Construct network
d = nx.DiGraph()

d.add_edge('Temperature', 'Proteobacteria')

d.add_edge('Humidity', 'Proteobacteria')
d.add_edge('Humidity', 'OC')

d.add_edge('Precipitation', 'Proteobacteria')
d.add_edge('Precipitation', 'OC')

d.add_edge('C3 Macrophytes', 'Proteobacteria')
d.add_edge('C3 Macrophytes', 'OC')

d.add_edge('C3 Mesophytes', 'Proteobacteria')

d.add_edge('C4 Megaphytes', 'Proteobacteria')

d.add_edge('OC', 'Proteobacteria')

d.add_edge('EC', 'Proteobacteria')
d.add_edge('EC', 'OC')

d.add_edge('pH', 'Proteobacteria')
d.add_edge('pH', 'OC')

# Assert DAG
assert nx.is_directed_acyclic_graph(d)

# d-separation

X = {'Humidity'}
Y = {'C3 Macrophytes', 'OC'}
Z = {'Temperature', 'Proteobacteria', 'Humidity'}

print(nx.d_separated(d, X, Y, Z))
```

From what I have read in [Molak 2023](https://www.amazon.com/dp/1804612987) and these [UBC CPSC notes](https://www.cs.ubc.ca/labs/lci/mlrg/slides/doCalc.pdf) is that d-separation is defined for disjoint sets. Unfortunately I don't have my [Pearl 2009](http://bayes.cs.ucla.edu/BOOK-2K/) right now.

I have also asked about non-disjointedness on SE: [Why is d-separation only for disjoint sets?](https://stats.stackexchange.com/questions/625390/why-is-d-separation-only-for-disjoint-sets)

So what `nx.d_separated` is doing might be a generalization of what (at least some of) the causal inference folks think of as checking d-separation.

So my question is what is `nx.d_separated` doing when the sets are non-disjoint?

**Comment by Alex-Markham** [comment, marked as answer]
Looking into the [code](https://github.com/networkx/networkx/blob/364e20eca8f270b9d364fab7f73ea63338790d98/networkx/algorithms/d_separation.py#L131), it seems the algorithm is:
- take the induced subgraph of G on the ancestor set of x union y union z
- remove all outgoing edges of nodes in z
- return True if no node in x is in the same weakly connected component as a node in y

So the answer is that the code is totally ignoring whether the sets are disjoint (which is a bug).

I would argue this isn't a meaningful generalization of d-separation (at least if you care about d-separation because of its relation to conditional independence) for the reasons Kuku gave in answer to your SE question.

Looking at this example also made some other bugs apparent. Adjacent nodes cannot be d-separated, but the current implementation says adjacent nodes are d-separated if you pick the right nondisjoint conditioning set (as you did in your example). Also `minimal_d_separator(D, 'a', 'b')` incorrectly returns `'c'` for the DAG D defined by a -> b <- c.

**Reply by Alex-Markham** [reply]
filed a bug report #6994
